lib/strategy/FragmentationStrategyShape.ts

```diff
@@ -0,0 +1,292 @@
+import { readFileSync } from 'fs';
+import { join } from 'path';
+import type * as RDF from '@rdfjs/types';
+import * as ShexParser from '@shexjs/parser';
+import { JsonLdParser } from 'jsonld-streaming-parser';
+import { DataFactory } from 'rdf-data-factory';
+import type { IQuadSink } from '../io/IQuadSink';
+import { FragmentationStrategyStreamAdapter } from './FragmentationStrategyStreamAdapter';
+import { FragmentationStrategySubject } from './FragmentationStrategySubject';
+// eslint-disable-next-line import/extensions
+import * as SHEX_CONTEXT from './shex_context.json';
+
+const DF = new DataFactory<RDF.Quad>();
+
+interface IShapeEntry {
+  /**
+   * @description the ShExC string
+   */
+  shape: string;
+  /**
+   * @description the directory targeted by the shape
+   */
+  directory: string;
+}
+
+export class FragmentationStrategyShape extends FragmentationStrategyStreamAdapter {
+  private readonly relativePath?: string;
+  private readonly tripleShapeTreeLocator?: boolean;
+  private readonly generationProbability?: number;
+  private readonly shapeMap: Map<string, IShapeEntry>;
+
+  private readonly irisHandled: Set<string> = new Set();
+  private readonly resourcesHandled: Set<string> = new Set();
```

Comment:
Do we need both of these sets? Or can we live with just one of them?

Reply:
We will have a lot of duplicate shape tree indicator if we don't use both. And if we don't use locator it will be a bit slower due to the for loop.

Reply:
But the places were you add to these sets are identical, which makes me think that the contents of these 2 sets will always be the same.

Reply:
They are not identical, I think, one check if the value is in `resourceHandled` and the other not. Because the idencator has to be in very file but the shape information is partition by resource.

Reply:
I think I see, I don't always capture the iri in `irihandled` and I should do it to avoid more time the for loop.